src/lib/trace/trace.go

```diff
@@ -0,0 +1,139 @@
+	var err error
+	var exp tracesdk.SpanExporter
+	cfg := GetGlobalConfig()
+	if len(cfg.Jaeger.Endpoint) != 0 {
+		// Jaeger collector exporter
+		log.Infof("init trace provider jaeger collector on %s with user %s", cfg.Jaeger.Endpoint, cfg.Jaeger.Username)
+		exp, err = jaeger.New(jaeger.WithCollectorEndpoint(
+			jaeger.WithEndpoint(cfg.Jaeger.Endpoint),
+			jaeger.WithUsername(cfg.Jaeger.Username),
+			jaeger.WithPassword(cfg.Jaeger.Password),
+		))
+	} else if len(cfg.Jaeger.AgentHost) != 0 {
+		// Jaeger agent exporter
+		log.Infof("init trace provider jaeger agent on %s", cfg.Jaeger.AgentHost)
+		exp, err = jaeger.New(jaeger.WithAgentEndpoint(
+			jaeger.WithAgentHost(cfg.Jaeger.AgentHost),
+			jaeger.WithAgentPort(cfg.Jaeger.AgentPort),
+		))
+	} else if len(cfg.Otel.Endpoint) != 0 {
+		// Otel exporter
+		log.Infof("init trace provider otel on %s/%s", cfg.Otel.Endpoint, cfg.Otel.URLPath)
+		opts := []otlptracehttp.Option{
+			otlptracehttp.WithEndpoint(cfg.Otel.Endpoint),
+			otlptracehttp.WithURLPath(cfg.Otel.URLPath),
+			otlptracehttp.WithTimeout(time.Duration(cfg.Otel.Timeout) * time.Second),
+		}
+		if cfg.Otel.Compression {
+			opts = append(opts, otlptracehttp.WithCompression(otlptracehttp.GzipCompression))
+		}
+		if cfg.Otel.Insecure {
+			opts = append(opts, otlptracehttp.WithInsecure())
+		}
+		exp, err = otlptracehttp.New(ctx, opts...)
+	} else {
+		log.Fatalf("Trace is enabled but no tracer provider is specified")
+	}
+	return exp, err
+}
+
+func initProvider(exp tracesdk.SpanExporter) (*tracesdk.TracerProvider, error) {
+	cfg := GetGlobalConfig()
+
+	// prepare attribute resources
+	attriSlice := []attribute.KeyValue{
+		semconv.ServiceNameKey.String(cfg.ServiceName),
+	}
+	if len(version.ReleaseVersion) != 0 {
+		attriSlice = append(attriSlice, semconv.ServiceVersionKey.String(version.ReleaseVersion))
+	}
+	if cfg.Namespace != "" {
+		attriSlice = append(attriSlice, semconv.ServiceNamespaceKey.String(cfg.Namespace))
+	}
+	if cfg.Attributes != nil {
+		for i, a := range cfg.Attributes {
+			attriSlice = append(attriSlice, attribute.String(i, a))
+		}
+	}
+
+	// prepare tp options
+	ops := make([]tracesdk.TracerProviderOption, 0, 4)
+	ops = append(ops,
+		// Always be sure to batch in production.
+		// tracesdk.WithBatcher(exp),
+		tracesdk.WithBatcher(exp),
+		// Record information about this application in an Resource.
+		tracesdk.WithResource(resource.NewWithAttributes(semconv.SchemaURL, attriSlice...)),
+		tracesdk.WithSampler(tracesdk.TraceIDRatioBased(cfg.SampleRate)),
+	)
+	// init trace provider
+	tp := tracesdk.NewTracerProvider(ops...)
+	return tp, nil
+}
+
+// ShutdownFunc is a function to shutdown the trace provider
+type ShutdownFunc func()
+
+// Shutdown shutdown the trace provider
+func (s ShutdownFunc) Shutdown() {
+	s()
+}
+
+// Init initializes the trace provider
+func InitGlobalTracer(ctx context.Context) ShutdownFunc {
+	if !Enabled() {
+		otel.SetTracerProvider(oteltrace.NewNoopTracerProvider())
```

Comment:
what about if we do not set the NoopTracerProvider when it's disabled? Seems no impact?

Reply:
yes, but we need to add tremendous code like `if trace.Enabled ...` to the place where you wanna add tracing